### PR TITLE
:seedling:  replacing DeepEqual with cmp Diff in tests

### LINKED ISSUE #8186: Consider replacing reflect.DeepEqual
I think we should consider replacing reflect.DeepEqual. Whilte it works well to diff two structs, it's very very had to debug what exactly is not equal. This makes it very hard to produce good logs to tell users why something was considered not equal.

I think go-cmp might be a good alternative. For more details, see: https://github.com/google/go-cmp

We already use go-cmp today to produce a diff in test code after reflect.DeepEqual only told us that something is not equal. The comparision algorithm of go-cmp seems to be better and additionally we get the beneift of getting a result which tells us where the difference is. This would be super valuable for our prod code to tell folks why we are doing things. 

Using cmp.Diff for logs while we use reflect.DeepEqual for the actual comparison is not great as there are small differences, e.g. unexported fields.

One example where I think this is super valuable is in the KCP controller where we could tell folks why exactly we are triggering a MachineRollout and not just: well reflect.DeepEqual told us that there is a difference.

There are a few more places like e.g. the ClusterClass controller where we compare variable definitions where we could have a log message (with high v) that could log exactly why the definitions are different.

(xref: https://github.com/kubernetes-sigs/cluster-api/pull/8107#discussion_r1109601666)

/kind feature

### PR DESCRIPTION
**Pull request by DiptoChakrabarty**
wrapper function to compare cmp diff in case panic occurs with log

**What this PR does / why we need it**:
replaces reflect.DeepEqual with go-cmp diff function and also creates a wrapper function to perform comparisons that handles panics which can be used for non testing code as suggested in issue discussion. Tests for the same written
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8186

### CODE REVIEW

File: exp/addons/api/v1alpha3/clusterresourcesetbinding_types_test.go [hunk 1]
```diff
@@ -147,7 +147,7 @@ func TestSetResourceBinding(t *testing.T) {
 			tt.resourceSetBinding.SetBinding(tt.resourceBinding)
 			exist := false
 			for _, b := range tt.resourceSetBinding.Resources {
-				if reflect.DeepEqual(b.ResourceRef, tt.resourceBinding.ResourceRef) {
+				if diff := cmp.Diff(b.ResourceRef, tt.resourceBinding.ResourceRef); diff == "" {
```
Comment (ykakarap) [inline exp/addons/api/v1alpha3/clusterresourcesetbinding_types_test.go:150]:
```suggestion
				if cmp.Equal(b.ResourceRef, tt.resourceBinding.ResourceRef) {
```

How about this?

(same in the other 2 clusterresourcesetbingins_types_test.go files)

File: util/conditions/compare.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"fmt"
+	"log"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func Check(actual interface{}, expected interface{}) (result bool, err error) {
+	defer func() error {
+		if recover() != nil {
+			log.Printf("panic occured got %v expected %v", actual, expected)
```
Comment (ykakarap) [inline util/conditions/compare.go:29]: I am not sure about the logs here. If the error needs to be logged it should be handled by the function using this util function.

File: util/conditions/compare.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+package conditions
+
+import (
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// Check compares the actual and expected values provided using the
```
Comment (enxebre) [inline util/conditions/compare.go:25]: I'm curious what's the motivation for this? are we unexpectedly panicking somewhere atm?

Comment (DiptoChakrabarty) [reply util/conditions/compare.go:25]: @enxebre I added this to catch panic cases if any occurs 

it was mentioned in the issue discussion

https://github.com/kubernetes-sigs/cluster-api/issues/8186#issuecomment-1467962898

**Comment by vincepri** [comment]
what's the status of this PR?

**Comment by sbueringer** [comment]
/close

Feel free to reopen if you want to continue this work

**Pull request closed without merging**